File: backend/app/esquemas/calendario_cosmico.py
"""Esquemas de respuesta para el Calendario Cósmico."""

from pydantic import BaseModel


class AspectoExactoRespuesta(BaseModel):
    """Aspecto exacto relevante del día."""

    planeta_a: str
    planeta_b: str
    tipo: str
    angulo: float
    orbe: float


class CambioSignoRespuesta(BaseModel):
    """Cambio de signo detectado respecto al día anterior."""

    planeta: str
    de: str
    a: str


class EventosCalendarioRespuesta(BaseModel):
    """Eventos notables calculados o persistidos para el día."""

    cambios_signo: list[CambioSignoRespuesta]
    retrogrados_inicio: list[str]
    retrogrados_fin: list[str]
    aspectos_exactos: list[AspectoExactoRespuesta]
    fases: str | None = None


class PlanetaCalendarioRespuesta(BaseModel):
    """Posición de un planeta en una fecha específica."""

    nombre: str
    longitud: float
    latitud: float
    signo: str
    grado_en_signo: float
    retrogrado: bool
    velocidad: float


class TransitosDiaRespuesta(BaseModel):
    """Tránsitos planetarios de un día específico."""

    fecha: str
    fecha_utc: str
    dia_juliano: float
    planetas: list[PlanetaCalendarioRespuesta]
    aspectos: list[AspectoExactoRespuesta]
    fase_lunar: str
    eventos: EventosCalendarioRespuesta
    estado: str


class CalendarioRangoRespuesta(BaseModel):
    """Tránsitos planetarios de un rango de fechas."""

    fecha_inicio: str
    fecha_fin: str
    dias: list[TransitosDiaRespuesta]

File: frontend/src/lib/tipos/calendario-cosmico.ts
/**
 * Tipos para el Calendario Cósmico de CosmicEngine.
 * Tránsitos planetarios por día y rango de fechas.
 */

/** Posición de un planeta en una fecha específica. */
export interface PlanetaCalendario {
  nombre: string;
  longitud: number;
  latitud: number;
  signo: string;
  grado_en_signo: number;
  retrogrado: boolean;
  velocidad: number;
}

export interface AspectoCalendario {
  planeta_a: string;
  planeta_b: string;
  tipo: string;
  angulo: number;
  orbe: number;
}

export interface CambioSignoCalendario {
  planeta: string;
  de: string;
  a: string;
}

export interface EventosCalendario {
  cambios_signo: CambioSignoCalendario[];
  retrogrados_inicio: string[];
  retrogrados_fin: string[];
  aspectos_exactos: AspectoCalendario[];
  fases: string | null;
}

/** Tránsitos planetarios de un día específico (mediodía UTC). */
export interface TransitosDia {
  fecha: string;
  fecha_utc: string;
  dia_juliano: number;
  planetas: PlanetaCalendario[];
  aspectos: AspectoCalendario[];
  fase_lunar: string;
  eventos: EventosCalendario;
  estado: "pasado" | "presente" | "futuro";
}

/** Tránsitos de un rango de fechas. */
export interface CalendarioRango {
  fecha_inicio: string;
  fecha_fin: string;
  dias: TransitosDia[];
}
